### ISSUE TITLE
[Discussion] Moving to python-bash hybrid based on Plugins

### ISSUE THREAD

**Issue by Ornias1993** [jailmanager/jailman#197]
Considering
- As discussed in #196 there are significant simplification gains if we use plugins
- As discussed in #196, https://github.com/iocage/iocage/issues/1193 and https://github.com/iocage/iocage/issues/1191 we still do need to work around some plugin limitations, at least before those changes are merged.
- As discussed in #171 there are significant code quality/readability gains in using python
- As discussed in #171 there are significant downsides to forcing people to use python for the blueprints
- As discussed with IXsystems/iocage, either plugins or hacks are required to run hardware transcoding
(https://jira.ixsystems.com/browse/NAS-105834) 
- We currently can't limit write access on a per-blueprint basis, this prevents us letting blueprint submitters to be "king of their own castle"
- Small updates on blueprints now need to follow the release schedule of the main plugin

I want to propose the following changes:
- Every blueprint would be turned into a plugin
- All copy operations will be moved into the plugin overlay structure where possible
- All PKG, versioning, repository and devfs settings will be moved to the plugin manifest
- Every plugin would also get a `/jailman` folder containing jailman post_install and upgrade steps
- After install and before update jailman would download the /jailman folder from the respective plugin repo
- The plugins (including jailman install/update scripts) will stay fully writen in BASH
- The main repo will start to accept PR's moving it to Python and will eventually aim to become python centric
- Plugin submitters/maintainers get more rights to manage their own plugins
- They do NOT have to limit themselves to release schedules of the main jailman repository

This discussion basically superseeds both #196 and #171
I'll leave this proposal open for feedback for a few days, if no one objects it would pretty much means this gets implemented.

People with a stake in this discussion:
@asmod3us @thorsteneb

**Comment by asmod3us** [comment]
Great initiative! I have other priorities right now so still working on digesting the deluge of updates :)
Quick questions:

1. is iocage open to the proposed changes at all? Seems we would depend on them getting implemented but the maintainers have not chimed in yet.
1. How would configuring jailman-specific add-ons to plugins work, e.g. traefik domain, providing datasets to jails?
1. How does this affect the user experience? Would jailman still work from the CLI, and/or be triggered by FreeNAS UI?

**Comment by Ornias1993** [comment]
1. The current proposal does not need those changes... Simply put: The plugins we will use for the time being would have a blank post_install script and we run the jailmanager install.sh on the plugin afterwards
2. We just use plugins as a streamlined way to install a customised jail. Nothing else. Basically the one line that installs the jail, would install the plugin instead. The plugin would just be a plain install of the jail, with the PKG's installed and the files copied. After plugin install jailman would run like normal (at least untill some changes in iocage are pushed).
3. Nothing would change for users of jailman, at all. The plugins are just a way for us to install things, they aren't "real" plugins, they are just iocage plugins (manifests, overlay and scripts) for us to streamline the deployment.

Simply put:
We currently: 
Turn a pkg list into a json -> install a jail based on those pkg's -> run jailman.

With our custom plugins we would do:
Install a jail based on a plugin manifest -> copy files from the overlay folder -> run jailman

In the FAR FAR future I would hope to get a UI in, but thats not something for now to even consider

**Comment by Ornias1993** [comment]
> digesting the deluge of updates :)

@asmod3us Don't worry too much about it, the changes since your last activity are rather simple:
- v1.2.3 is merged into master as-is, but everything was tested and some slight fixes where applied (nothing noteworthy)
- Dev was rebased on the newest master release
- I merged the previous PR with traefik stuff as-is into dev (1.3.0) (because dev needed complete retesting after the rebase anyway)
- I'm currently testing on TrueNAS CORE Beta and fixing directly into the dev branch. (again no noteworthy fixes as of yet)

**Comment by Ornias1993** [comment]
I wrote a short white paper to explain the proposed changes:

[Project Structure Whitepaper.md.txt](https://github.com/jailmanager/jailman/files/4938112/Project.Structure.Whitepaper.md.txt)

**Comment by thorsteneb** [comment]
Thank you. This is basically a CONTRIBUTING.md, or part of one, do I get that right?

I have a question about jails that rely on host-side changes. I think I know the answer, and want to make sure.

1. Jailman starts, checks if the plugin exists, downloads the plugin, executes general tests and starts the plugin installer
2. Plugin install (installs a jail with the specified PKGS)
3. Plugin `post_install.sh` (runs the `post_install.sh` script inside(!) the jail)
4. Jailman `/jailman/finish_install.sh` (runs `finish_install.sh` on the host(!))

If a plugin requires a driver, for example Plex with hardware transcode, it would output (via echo) instructions to the user to create the relevant tunable, since it cannot pass instructions to jailman in step 4) to create that tunable for the user. Do I have that right, or am I misunderstanding?

I do think that leaving the creation of tunables to the user is probably best, as loading drivers can crash TrueNAS for certain combinations of hardware.

**Comment by Ornias1993** [comment]
@thorsteneb That was something I was playing with a LOT in my headspace...
Because that issue goes byond just drivers, just mounting the /config directory already needs host-access or connecting a plugin to traefik.

The way I solved this is actually pretty simple:
`/jailman/finish_install.sh` is part of the plugin repository. Its basically the current install.sh in the blueprint directory, nothing more, nothing less.

**Comment by thorsteneb** [comment]
> `/jailman/finish_install.sh` is part of the plugin repository.

Ah, thank you for explaining that! That makes sense. You do call that out in the whitepaper, I missed it.

**Comment by Ornias1993** [comment]
Okey, I think everyone somewhat agree's, so i'm preparing a Proof of Concept as we speak.
Because me and @thorsteneb where also looking into hardware passthrough before, i''m going to make Plex the target of my PoC... Keep you guys in touch! :)

**Comment by Ornias1993** [comment]
The fetching and installing based on plugins is mostly done for the PoC...
Setting the right devfs rules for plex also works.

Currently working on integrating the variables a little better.

**Comment by Ornias1993** [comment]
Proof of Concept is in... it has become (how appropriate) #200 
It think that suffices to say this discussion can be considered closed and moved to the "design review" stage...
